Restrict upload quota according to plan

hackerkid opened the pull request.

timabbott commented on zerver/models.py:
```diff
@@ -397,6 +397,12 @@ def upload_quota_bytes(self) -> Optional[int]:
         # it as gibibytes (GiB) to be a bit more generous in case of confusion.
         return self.upload_quota_gb << 30
 
+    def currently_used_upload_space_bytes(self) -> int:
```
We should probably make this function have a cache that lasts an hour, so that it doesn't become a significant performance bottleneck.

hackerkid commented:
@timabbott @rishig This is ready for a review :)

timabbott commented:
I don't like the approach of sending an event to every user in the realm every time someone uploads a file, since that seems like a huge amount of wasted work for a large organization.   Is that just to show the quota status in the "attachments" UI?  I think it'd be better to just fetch that data from the server when we render that settings UI page.

hackerkid commented:
@timabbott Pinging again in case you missed the previous notification :)

timabbott commented on static/js/attachments_ui.js:
```diff
@@ -33,9 +19,45 @@ function bytes_to_size(bytes, kb_with_1024_bytes) {
+    if (page_params.realm_upload_quota === null) {
+        return null;
+    }
+    return (uploads_size / page_params.realm_upload_quota).toFixed(1);
```
This is missing a `100 *` to get percentage, and thus always outputs 0.0%.  I found this by testing uploading a 3MB file after setting the quota to 1GB.  You should definitely has tested this manually. :(

hackerkid replied:
@timabbott Sorry. I actually tested this a lot of times. I was thinking it was showing 0% due to the small upload space used. So I was returning just `uploads_size` during the testing to make sure that upload_size is getting updated during events.

hackerkid replied:
Attaching a screenshot I did during the testing phase to make sure that the percentage changes when I upload a lot of big files. It should have shown 10% instead of 0.1% but i did this.
https://chat.zulip.org/user_uploads/2/dd/FHcAP7QFoHyqdnG890EcTk-6/Screenshot-from-2019-01-15-13-06-57.png

hackerkid replied:
Attaching a screenshot I took during the testing phase to make sure that the percentage changes when I upload a lot of big files. It should have shown 10% instead of 0.1% but i did test this a lot. I was not super careful with math. But I did test this a lot. 
https://chat.zulip.org/user_uploads/2/dd/FHcAP7QFoHyqdnG890EcTk-6/Screenshot-from-2019-01-15-13-06-57.png

hackerkid replied:
Attaching a screenshot I took during the testing phase to make sure that the percentage changes when I upload a lot of big files. It should have shown 10% instead of 0.1% but i did test this a lot. I was not careful with math which was a mistake.
https://chat.zulip.org/user_uploads/2/dd/FHcAP7QFoHyqdnG890EcTk-6/Screenshot-from-2019-01-15-13-06-57.png

timabbott commented on static/js/attachments_ui.js:
```diff
@@ -33,9 +19,45 @@ function bytes_to_size(bytes, kb_with_1024_bytes) {
+            var percent_used = exports.percentage_used_space(data.upload_space_used);
+            var upload_quota = attachments_ui.bytes_to_size(page_params.realm_upload_quota, true);
+            $("#upload_percentage_used").text(percent_used);
+            $("#upload_quota").text(upload_quota);
```
So, the problem with this approach is that there's a brief glitch where the data isn't displayed yet.  We should instead have a small handlebars template for this widget, and just have the whole element added by this code path.

timabbott replied:
Alternatively, there's a `channel.get` call that we already do before rendering this UI (see below); maybe we should just add the `upload_space_used` value to the dictionary returned there (removing the new endpoint) and remove this extra call?  That might be cleaner.

timabbott commented:
@hackerkid thanks for working on this!  Overall, this looks great, and I merged all but the last commit.  I posted comments on the last one (and also pushed back to it to fix the math bug I noted).

It's possible based on the resolution to my comments above that we'll want to remove the second-to-last commit (the one that adds a new API endpoint) too as part of restructuring this to avoid that glitch, but that's OK, since the more I think about this the more it seems like we should just fold this into the existing "attachments" endpoint.

Finally, we should remember to do a manual migration of `upload_quota_gb` on existing organizations when we roll this out; otherwise those organizations will never see this alert as their `upload_quota_gb` will always be `None`.

hackerkid commented:
@timabbott I have removed the separate endpoint for fetching upload_space_used. Instead of that I am now passing the info via /attachments endpoint as you suggested. I have also updated the attachment  event system to include upload_space used so that the uploaded files UI would be updated.

![screenshot from 2019-03-07 22-06-15](https://user-images.githubusercontent.com/7190633/53975687-77c04800-412b-11e9-9547-de6e94ec4157.png)

timabbott commented:
Nice, merged, thanks @hackerkid!  @rishig FYI.
